bioconda_utils/utils.py

```diff
@@ -721,115 +771,88 @@ def changed_since_master(recipe_folder):
-
-    channels : None or list
-        Optional list of channels to check for existing recipes
-
-    force : bool
-        Build the package even if it is already available in supplied channels.
-    """
-    if not isinstance(env_matrix, EnvMatrix):
-        env_matrix = EnvMatrix(env_matrix)
-
+def get_package_paths(recipe, channel_packages, force=False):
+    # check if package is noarch, if so, build only on linux
+    # with temp_os, we can fool the MetaData if needed.
+    platform = os.environ.get('OSTYPE', sys.platform)
+    if platform.startswith("darwin"):
+        platform = 'osx'
+    elif platform == "linux-gnu":
+        platform = "linux"
+
+    metas = load_all_meta(
+        recipe, config=load_conda_build_config(platform=platform))
+
+    # The recipe likely defined skip: True
+    if not metas:
+        return []
+
+    # If on CI, handle noarch.
+    if os.environ.get('CI', None) == 'true':
+        first_meta = metas[0]
+        if first_meta.get_value('build/noarch'):
+            if platform != 'linux':
+                logger.debug('FILTER: only building %s on '
+                             'linux because it defines noarch.',
+                             recipe)
+                return []
+
+    new_metas, existing_metas, divergent_builds = (
+        _filter_existing_packages(metas, channel_packages))
+
+    if divergent_builds:
+        raise DivergentBuildsError(*sorted(divergent_builds))
+
+    for meta in existing_metas:
+        logger.info(
+            'FILTER: not building %s because '
+            'it is in channel(s) and it is not forced.', meta.pkg_fn())
+    # yield all pkgs that do not yet exist
+    if force:
+        build_metas = new_metas + existing_metas
+    else:
+        build_metas = new_metas
+    return list(chain.from_iterable(
+        api.get_output_file_paths(meta) for meta in build_metas))
+
+
+def _filter_existing_packages(metas, channel_packages):
+    new_metas = []  # MetaData instances of packages not yet in channel
+    existing_metas = []  # MetaData instances of packages already in channel
+    divergent_builds = set()  # set of Dist (i.e., name-version-build) strings
+
+    recipe_pkgs = defaultdict(list)
+    for meta in metas:
+        pkg_key = PackageKey(
+            meta.name(), meta.version(), int(meta.build_number() or 0))
+        recipe_pkgs[pkg_key].append(meta)
+    recipe_pkgs.default_factory = None
+
+    for pkg_key, pkg_metas in recipe_pkgs.items():
+        existing_build_ids = channel_packages.get(pkg_key, set())
+        for meta in pkg_metas:
+            if meta.build_id() not in existing_build_ids:
+                new_metas.append(meta)
+            else:
+                existing_metas.append(meta)
+        pkg_build_ids = set(meta.build_id() for meta in pkg_metas)
+        for divergent_build in (existing_build_ids - pkg_build_ids):
+            divergent_builds.add(
+                '-'.join((pkg_key.name, pkg_key.version, divergent_build)))
```

Comment:
TODO/FIXME: This misses a corner case where packages have mismatches *only* in the `osx-64` subdir since `channel_packages` only contains data from `linux-64` and `noarch` when we have a `noarch` build.
Solution could be to always fail when we do a `noarch` build and there are builds with the same number in `linux-64` or `osx-64`.
Otherwise we'd have to also render with `CONDA_SUBDIR=osx-64` (at least when we do a `noarch` build) to be able to inspect those build strings. But we don't have to go through this trouble since we do not support the case where we have builds with the same number in `noarch` and simultaneously other subdirs <sup>(That's actually valid, albeit the use cases are rather exotic, currently.)</sup>.